packages/app-check/src/internal-api.ts

```diff
@@ -209,14 +167,23 @@ export function addTokenListener(
     newState.tokenRefresher.start();
   }
 
-  // invoke the listener async immediately if there is a valid token
+  // Invoke the listener async immediately if there is a valid token
+  // in memory.
   if (state.token && isValid(state.token)) {
     const validToken = state.token;
     Promise.resolve()
       .then(() => listener({ token: validToken.token }))
       .catch(() => {
         /** Ignore errors in listeners. */
       });
+  } else {
+    // Also try storage. Otherwise isTokenAutoRefreshEnabled == false
+    // will prevent reading existing token from storage.
+    void readTokenFromStorage(app).then(cachedToken => {
```

Comment:
I realized that without this, listeners won't read tokens in storage if automatic token refresh is disabled, and I don't think that's expected behavior. If the refresher doesn't start, getToken isn't called, and that's the only place that reads from storage. I'm not sure if this is the best place to change it or if we need to rethink where we block token refresh.

Reply:
This is a little complicated, that means either:
1. activate() needs to be made async, which I don't think is an option
2. the in-memory token needs to always be stored as a promise (maybe?)
3. leave it a hanging promise, if it didn't manage to get the token from storage in time for the next getToken() call, forget it and try the exchange endpoint. Probably not ok?

Reply:
> 1. activate() needs to be made async, which I don't think is an option
> 2. the in-memory token needs to always be stored as a promise (maybe?)
> 3. leave it a hanging promise, if it didn't manage to get the token from storage in time for the next getToken() call, forget it and try the exchange endpoint. Probably not ok?

Yeah, we will need to store a promise for reading the cached token in the state. Here is the workflow when we need a token:
- if `state.token` is defined and valid, return it
- if `state.token` is defined and invalid, try the exchange endpoint
- If `state.token` is undefined, wait for this promise to resolve
- if the promise resolves to null (no cached token), or an invalid token, try the exchange endpoint

The benefits of doing it are that `readTokenFromStorage` is only called in a single place and it won't be called the second time unnecessarily, which is possible with the current implementation. The downside is that we need to manage one additional variable in the state. I will leave it up to you to decide.

Reply:
I missed this comment in the last review. Addressed now.